contributors/design-proposals/sig-storage/volume-topology-scheduling.md

```diff
@@ -0,0 +1,580 @@
+    * With this new behavior, the PVC will not rebind because binding would only
+be triggered when the Pod has to go through scheduling.  This could cause
+problems with Pod volume cleanup because the PVC will no longer be bound.
+    * But deleting the PVC and PV (especially both) while a Pod is still using it
+is also dangerous and should be prevented in the first place.  There could be
+serious problems if you recreate the PVC and it rebinds to a different PV.  If
+the PV had a reclaim policy of Delete, then deleting the PVC could cause the
+underlying disk to get deleted while the Pod is still using it.
+
+
+#### PV Controller Changes
+When the feature gate is enabled, the PV controller needs to skip binding and
+provisioning all unbound PVCs that don’t have prebound PVs, and let it come
+through the scheduler path.  This is the block in `syncUnboundClaim` where
+`claim.Spec.VolumeName == “”` and there is no PV that is prebound to the PVC.
+
+No other state machine changes are required.  The PV controller continues to
+handle the remaining scenarios without any change.
+
+The methods to find matching PVs for a claim, prebind PVs, and to invoke and
+handle dynamic provisioning need to be refactored for use by the new scheduler
+functions.
+
+#### Scheduler Changes
+
+##### Predicate
+A new predicate function checks all of a Pod's unbound PVCs can be satisfied
+by existing PVs or dynamically provisioned PVs that are
+topologically-constrained to the Node.
+```
+MatchUnboundPVCs(pod *v1.Pod, node *v1.Node) (canBeBound bool, err error)
+```
+1. If all the Pod’s PVCs are bound, return true.
+2. Otherwise try to find matching PVs for all of the unbound PVCs in order of
+decreasing requested capacity.
+3. Walk through all the PVs.
+4. Find best matching PV for the PVC where PV topology is satisfied by the Node.
+5. Temporarily cache this PV in the PVC object, keyed by Node, for fast
+processing later in the priority and bind functions.
+6. Return true if all PVCs are matched.
+7. If there are still unmatched PVCs, check if dynamic provisioning is possible.
+For this alpha phase, the provisioner is not topology aware, so the predicate
+will just return true if there is a provisioner specified (internal or external).
+8. Otherwise return false.
+
+TODO: caching format and details
+
+##### Priority
+After all the predicates run, there is a reduced set of Nodes that can fit a
+Pod. A new priority function will rank the remaining nodes based on the
+unbound PVCs and their matching PVs.
+```
+PrioritizeUnboundPVCs(pod *v1.Pod, filteredNodes HostPriorityList) (rankedNodes HostPriorityList, err error) 
+```
+1. For each Node, get the cached PV matches for the Pod’s PVCs.
+2. Compute a priority score for the Node using the following factors:
+    1. How close the PVC’s requested capacity and PV’s capacity are.
+    2. Matching static PVs is preferred over dynamic provisioning.
```

Comment:
Why are static PVs preferred over dynamic?

Reply:
It's assuming that you had specifically statically created the PVs for your application.